Notifications Service PR

skuzmanoski opened the pull request.

skuzmanoski committed:
Add initial project structure with some configuration.

skuzmanoski committed:
Add Kafka for messaging.

skuzmanoski committed:
Create README.md

skuzmanoski committed:
Update README.md

skuzmanoski committed:
Update docker-compose.yml

Removing Kafka Manager from comments.

stefankochev commented on notifications-service/docker-compose.yml:
```diff
@@ -0,0 +1,51 @@
+version: '3.8'
+
+services:
+  zookeeper:
```
@SKuzmanoski and team, let's move the docker-compose file to project's root (root dir), then reference the `notifications-service` build with `build: notifications-service` instead of `build: .`, because all teams should use the same docker-compose, otherwise the scaffold looks ok ✅

stefankochev replied:
also I'm getting the following error:
```
 => CACHED [2/3] WORKDIR /app                                                                                                                                    0.0s
 => ERROR [3/3] COPY target/notifications-service-0.0.1.jar .                                                                                                    0.0s
------
 > [3/3] COPY target/notifications-service-0.0.1.jar .:
------
failed to solve: failed to compute cache key: "/target/notifications-service-0.0.1.jar" not found: not found
```

try implementing a multi stage Dockerfile
https://docs.docker.com/build/building/multi-stage/
where the first stage will produce the .jar and the second stage will copy and run the jar as you have it currently.
here is some [example](https://www.alibabacloud.com/help/en/container-registry/latest/build-an-image-for-a-java-application-by-using-a-dockerfile-with-multi-stage-builds)

skuzmanoski replied:
It should work now, thank you for the provided reference.

stefankochev replied:
Works now, thanks!

skuzmanoski committed:
Fix Dockerfile and Move docker-compose.yml.

Fixing Dockerfile so that it takes the whole project in the container and then builds the .jar also moved docker-compose.yml to be on whole project level.

vladimirstojcheski committed:
mail scheduler implemented and model updated

stefankochev commented on notifications-service/src/main/java/mk/ukim/finki/notifications_service/service/job/MailScheduler.java:
```diff
@@ -0,0 +1,47 @@
+    @Autowired
+    private PostNotificationService notificationService;
+
+    @Scheduled(cron = "0 */1 * ? * *")
```
Team, eventually the service should be configured to run on multiple instances, we shouldn't have internal cron scheduling because every instance of the service will run the independent schedule right?! And if the service logic is not dealing with this, we'll end up with multiple notifications when we need one... So we maybe need an external cron service, as we had for labs?! https://github.com/stefankochev/soa-labs/pull/4, check the diagram please.

stefankochev commented:
Merging this PR so we have a scaffold the other teams can open PRs and build on top of. Please keep sending the updates in next PRs 🙏
